src/System/Console/Args/Generics.hs

```diff
@@ -11,80 +12,163 @@
+    (options, arguments, parseErrors) ->
+      let result :: Either [String] (NP I xs) =
+            collectErrors $ project options (mkEmptyArguments fields)
+          allErrors =
+            parseErrors ++
+            map mkUnknownArgumentError arguments ++
+            ignoreRight result
+      in case allErrors of
+        [] -> result
+        _ -> Left allErrors
+  where
+    mkUnknownArgumentError :: String -> String
+    mkUnknownArgumentError arg = "unknown argument: " ++ arg
+
+    ignoreRight :: Monoid e => Either e o -> e
+    ignoreRight = either id (const mempty)
+
+mkOptDescrs :: forall xs . All Option xs =>
+  NP FieldInfo xs -> [OptDescr (NS (Either String) xs)]
+mkOptDescrs fields =
+  map toOptDescr $ sumList $ npMap mkOptDescr fields
+
+newtype OptDescrE a = OptDescrE (OptDescr (Either String a))
+  deriving (Functor)
+
+mkOptDescr :: forall a . Option a => FieldInfo a -> OptDescrE a
+mkOptDescr (FieldInfo name) = OptDescrE $ Option [] [name] toOption ""
+
+toOptDescr :: NS OptDescrE xs -> OptDescr (NS (Either String) xs)
+toOptDescr (Z (OptDescrE a)) = fmap Z a
+toOptDescr (S a) = fmap S (toOptDescr a)
+
+mkEmptyArguments :: forall xs . (SingI xs, All Option xs) =>
+  NP FieldInfo xs -> NP (Either String) xs
+mkEmptyArguments fields = case (sing :: Sing xs, fields) of
+  (SNil, Nil) -> Nil
+  (SCons, FieldInfo name :* r) ->
+    emptyOption name :* mkEmptyArguments r
+  _ -> error "mkEmpty: impossible"
+
+
+-- * showing help?
+
+helpWrapper :: (All Option xs) =>
+  [String] -> NP FieldInfo xs -> a -> Either (IO ()) a
+helpWrapper args fields a =
+    case getOpt Permute [helpOption] args of
+      ([], _, _) -> Right a
+      (() : _, _, _) -> Left $ do
+        progName <- getProgName
+        let header = progName
+        putStrLn (usageInfo header (mkOptDescrs fields))
   where
-    err :: String -> Either String (Parser a)
-    err message =
-      Left ("args-generics doesn't support " ++ message ++ " (" ++ typeName ++ ").")
+    helpOption = Option ['h'] ["help"] (NoArg ()) "show help and exit"
+
+
+-- * helper functions for NS and NP
+
+collectErrors :: NP (Either String) xs -> Either [String] (NP I xs)
+collectErrors np = case np of
+  Nil -> Right Nil
+  (a :* r) -> case (a, collectErrors r) of
+    (Right a, Right r) -> Right (I a :* r)
+    (Left error, r) -> Left (error : either id (const []) r)
+    (Right _, Left errors) -> Left errors
+
+npMap :: (All Option xs) => (forall a . Option a => f a -> g a) -> NP f xs -> NP g xs
+npMap _ Nil = Nil
+npMap f (a :* r) = f a :* npMap f r
+
+sumList :: NP f xs -> [NS f xs]
+sumList Nil = []
+sumList (a :* r) = Z a : map S (sumList r)
+
+project :: SingI xs => [NS f xs] -> NP f xs -> NP f xs
+project sums empty =
+    foldl' inner empty sums
+  where
+    inner :: NP f xs -> NS f xs -> NP f xs
+    inner (_ :* r) (Z a) = a :* r
+    inner (a :* r) (S rSum) = a :* inner r rSum
+    inner Nil _ = error "project: impossible"
```

Comment:
Again, module name please.  If it turns out that there are so many places where we need this (there is one more in this PR), the maybe abstract it:

```haskell
impossible :: String -> a
impossible name = "System.Console.Args.Generics." ++ name ++ ": This should never happen!"
```